src/subwords.rs

```diff
@@ -0,0 +1,42 @@
+use finalfusion::subword::{NGramsIndices, SubwordIndices};
+use pyo3::prelude::*;
+use pyo3::types::PyTuple;
+
+#[pyclass(name=SuwbwordInfo)]
+pub struct PySubwordInfo {
+    word: String,
+}
+
+#[pymethods]
+impl PySubwordInfo {
```

Comment:
I have been implementing this with all three approaches.

When I was writing it as a standalone function, I found that I need to put the function into `lib.rs` as a static function so that I can use the attribute `#[pyfunction]` and register it in module via `wrap_pyfunction!`. Which seems a bit strange... I don't know if there is anyway we can register this function in the module while keeping it in `subword.rs`

And about the approach: hash parameters as data structures, it went well.

Reply:
I gave this some more thought and I have also implemented support for fastText subword vocabs. I now think the best way to implement this is as a method of the Python `Vocab` class. This has various benefits:

* The `min_n`, `max_n`, and `bucket_exp` parameters from the vocabulary can be used.
* Once the `Indexer` is merged in `finalfusion-rust`, it can automatically use the correct hash function (in fact, the implementation does not need to know anything about `min_n`, `max_n`, and `bucket_exp`.
* The indices would be correct offsets into the embedding matrix (the vocab size is added to the indices).

One complication is that `SimpleVocab` does not provide subword indexing. So, we'd need appropriate traits in `finalfusion-rust`.

You may want to wait updating this PR until `Indexer`, `BucketIndexer`, and the parametrized `SubwordVocab` are merged in `finalfusion-rust`. Otherwise, you'd need to rewrite the code anyway once these PRs are in.

Reply:
Thanks!! This make so much more sense!

Reply:
I have been thinking for days and I'm still a bit confused here...

> * The `min_n`, `max_n`, and `bucket_exp` parameters from the vocabulary can be used.

I tried to do it but `min_n`, `max_n` and `bucket_exp` are private in vocabulary, unless we write a getter to get these values.

> in fact, the implementation does not need to know anything about `min_n`, `max_n`, and `bucket_exp`, since these are implementation details of an `Indexer`

the `Indexer` of a vocabulary is also private, I guess we would have to need a getter too

Reply:
No, we should avoid exposing to many internals. That `SubwordVocab` uses an indexer is an implementation detail and that it brackets the words and adds the vocab size to the subword indices are other implementation details.

The following PR should provide what is necessary for implementing this in `finalfusion-python`.

finalfusion/finalfusion-rust#42

Should be ready for merging pretty quickly once Sebastian is back.